### ISSUE TITLE
win x32 build is failing

### ISSUE THREAD

**Issue by kcpevey** [Quansight/numpy-threading-extensions#44]
In [this action](https://github.com/Quansight/numpy-threading-extensions/actions/runs/329943694), the win x32 build is failing with traceback:

```
python -m cibuildwheel --output-dir wheelhouse
  shell: C:\Program Files\Git\bin\bash.EXE --noprofile --norc -e -o pipefail {0}
  env:
    pythonLocation: C:\hostedtoolcache\windows\Python\3.7.9\x64
    CIBW_BUILD: cp36-win32

     _ _       _ _   _       _           _
 ___|_| |_ _ _|_| |_| |_ _ _| |_ ___ ___| |
|  _| | . | | | | | . | | | |   | -_| -_| |
|___|_|___|___|_|_|___|_____|_|_|___|___|_|

cibuildwheel version 1.6.3

Build options:
  platform: 'windows'
  before_all: ''
  before_build: None
  before_test: None
  build_selector: BuildSelector('cp36-win32' - '')
  build_verbosity: 0
  dependency_constraints: <cibuildwheel.util.DependencyConstraints object at 0x0000013991FA0348>
  environment: ParsedEnvironment([])
  manylinux_images: None
  output_dir: WindowsPath('wheelhouse')
  package_dir: WindowsPath('.')
  repair_command: ''
  test_command: None
  test_extras: ''
  test_requires: []

Here we go!

+ Download https://dist.nuget.org/win-x86-commandline/latest/nuget.exe to C:\cibw\nuget.exe
+ C:\cibw\nuget.exe install pythonx86 -Version 3.6.8 -OutputDirectory C:\cibw\python
Feeds used:
  https://api.nuget.org/v3/index.json
  C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\



Attempting to gather dependency information for package 'pythonx86.3.6.8' with respect to project 'C:\cibw\python', targeting 'Any,Version=v0.0'
Gathering dependency information took 1.96 sec
Attempting to resolve dependencies for package 'pythonx86.3.6.8' with DependencyBehavior 'Lowest'
Resolving dependency information took 0 ms
Resolving actions to install package 'pythonx86.3.6.8'
Resolved actions to install package 'pythonx86.3.6.8'
Retrieving package 'pythonx86 3.6.8' from 'nuget.org'.
  GET https://api.nuget.org/v3-flatcontainer/pythonx86/3.6.8/pythonx86.3.6.8.nupkg
  OK https://api.nuget.org/v3-flatcontainer/pythonx86/3.6.8/pythonx86.3.6.8.nupkg 7ms
  Building wheel for fast-numpy-loops (PEP 517): started
  ERROR: Command errored out with exit status 1:
  Building wheel for fast-numpy-loops (PEP 517): finished with status 'error'
   command: 'c:\cibw\python\pythonx86.3.6.8\tools\python.exe' 'c:\cibw\python\pythonx86.3.6.8\tools\lib\site-packages\pip\_vendor\pep517\_in_process.py' build_wheel 'C:\Users\RUNNER~1\AppData\Local\Temp\tmpaymsc_bq'
       cwd: C:\Users\RUNNER~1\AppData\Local\Temp\pip-req-build-uasvqgx9
  Complete output (81 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win32-3.6
  creating build\lib.win32-3.6\fast_numpy_loops
  copying src\fast_numpy_loops\__init__.py -> build\lib.win32-3.6\fast_numpy_loops
  creating build\lib.win32-3.6\fast_numpy_loops\tests
  copying src\fast_numpy_loops\tests\__init__.py -> build\lib.win32-3.6\fast_numpy_loops\tests
  running egg_info
  writing src\fast_numpy_loops.egg-info\PKG-INFO
  writing dependency_links to src\fast_numpy_loops.egg-info\dependency_links.txt
  writing requirements to src\fast_numpy_loops.egg-info\requires.txt
  writing top-level names to src\fast_numpy_loops.egg-info\top_level.txt
  reading manifest file 'src\fast_numpy_loops.egg-info\SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  warning: no directories found matching 'docs'
  warning: no files found matching '.appveyor.yml'
  warning: no previously-included files matching '*.py[cod]' found anywhere in distribution
  warning: no previously-included files matching '__pycache__\*' found anywhere in distribution
  warning: no previously-included files matching '*.so' found anywhere in distribution
  warning: no previously-included files matching '*.dylib' found anywhere in distribution
  writing manifest file 'src\fast_numpy_loops.egg-info\SOURCES.txt'
  copying src\fast_numpy_loops\_fast_numpy_loops.cpp -> build\lib.win32-3.6\fast_numpy_loops
  copying src\fast_numpy_loops\module_init.cpp -> build\lib.win32-3.6\fast_numpy_loops
  copying src\fast_numpy_loops\recycler.cpp -> build\lib.win32-3.6\fast_numpy_loops
  copying src\fast_numpy_loops\stubs.h -> build\lib.win32-3.6\fast_numpy_loops
  running build_ext
  building 'fast_numpy_loops._fast_numpy_loops' extension
  creating build\temp.win32-3.6
  creating build\temp.win32-3.6\Release
  creating build\temp.win32-3.6\Release\src
  creating build\temp.win32-3.6\Release\src\fast_numpy_loops
  creating build\temp.win32-3.6\Release\src\atop
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\bin\HostX86\x86\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Isrc/fast_numpy_loops -Isrc/atop -IC:\Users\RUNNER~1\AppData\Local\Temp\pip-build-env-ab3mlyvn\overlay\Lib\site-packages\numpy\core\include -Ic:\cibw\python\pythonx86.3.6.8\tools\include -Ic:\cibw\python\pythonx86.3.6.8\tools\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt" /EHsc /Tpsrc/fast_numpy_loops/_fast_numpy_loops.cpp /Fobuild\temp.win32-3.6\Release\src/fast_numpy_loops/_fast_numpy_loops.obj -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -DPy_LIMITED_API /Ox /Ob2 /Oi /Ot /d2FH4-
  _fast_numpy_loops.cpp
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\atop\common_inc.h(136): error C2371: 'INT_PTR': redefinition; different basic types
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(134): note: see declaration of 'INT_PTR'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\atop\common_inc.h(136): error C2371: 'PINT_PTR': redefinition; different basic types
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(134): note: see declaration of 'PINT_PTR'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\atop\common_inc.h(137): error C2371: 'UINT_PTR': redefinition; different basic types
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(135): note: see declaration of 'UINT_PTR'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\atop\common_inc.h(137): error C2371: 'PUINT_PTR': redefinition; different basic types
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(135): note: see declaration of 'PUINT_PTR'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\atop\common_inc.h(139): error C2371: 'LONG_PTR': redefinition; different basic types
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(137): note: see declaration of 'LONG_PTR'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\atop\common_inc.h(139): error C2371: 'PLONG_PTR': redefinition; different basic types
Failed to build fast-numpy-loops
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(137): note: see declaration of 'PLONG_PTR'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\atop\common_inc.h(140): error C2371: 'ULONG_PTR': redefinition; different basic types
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(138): note: see declaration of 'ULONG_PTR'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\atop\common_inc.h(140): error C2371: 'PULONG_PTR': redefinition; different basic types
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(138): note: see declaration of 'PULONG_PTR'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\atop\common_inc.h(158): warning C4005: 'InterlockedDecrement64': macro redefinition
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(7235): note: see previous definition of 'InterlockedDecrement64'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\atop\common_inc.h(159): warning C4005: 'InterlockedIncrement64': macro redefinition
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(7213): note: see previous definition of 'InterlockedIncrement64'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\fast_numpy_loops\../atop/threads.h(94): error C2371: 'LONG_PTR': redefinition; different basic types
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(137): note: see declaration of 'LONG_PTR'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\fast_numpy_loops\../atop/threads.h(94): error C2371: 'PLONG_PTR': redefinition; different basic types
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(137): note: see declaration of 'PLONG_PTR'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\fast_numpy_loops\../atop/threads.h(95): error C2371: 'ULONG_PTR': redefinition; different basic types
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(138): note: see declaration of 'ULONG_PTR'
  C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-uasvqgx9\src\fast_numpy_loops\../atop/threads.h(95): error C2371: 'PULONG_PTR': redefinition; different basic types
  C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\basetsd.h(138): note: see declaration of 'PULONG_PTR'
  src/fast_numpy_loops/_fast_numpy_loops.cpp(267): warning C4244: '=': conversion from 'int64_t' to 'npy_intp', possible loss of data
  src/fast_numpy_loops/_fast_numpy_loops.cpp(270): warning C4244: '=': conversion from 'int64_t' to 'npy_intp', possible loss of data
  src/fast_numpy_loops/_fast_numpy_loops.cpp(358): warning C4838: conversion from 'int64_t' to 'npy_intp' requires a narrowing conversion
  src/fast_numpy_loops/_fast_numpy_loops.cpp(358): warning C4244: 'initializing': conversion from 'int64_t' to 'npy_intp', possible loss of data
  src/fast_numpy_loops/_fast_numpy_loops.cpp(358): warning C4838: conversion from 'int64_t' to 'npy_intp' requires a narrowing conversion
  src/fast_numpy_loops/_fast_numpy_loops.cpp(358): warning C4244: 'initializing': conversion from 'int64_t' to 'npy_intp', possible loss of data
  src/fast_numpy_loops/_fast_numpy_loops.cpp(358): warning C4838: conversion from 'int64_t' to 'npy_intp' requires a narrowing conversion
  src/fast_numpy_loops/_fast_numpy_loops.cpp(358): warning C4244: 'initializing': conversion from 'int64_t' to 'npy_intp', possible loss of data
  src/fast_numpy_loops/_fast_numpy_loops.cpp(395): warning C4838: conversion from 'const int64_t' to 'const npy_intp' requires a narrowing conversion
  src/fast_numpy_loops/_fast_numpy_loops.cpp(395): warning C4244: 'initializing': conversion from 'const int64_t' to 'const npy_intp', possible loss of data
  src/fast_numpy_loops/_fast_numpy_loops.cpp(395): warning C4838: conversion from 'const int64_t' to 'const npy_intp' requires a narrowing conversion
  src/fast_numpy_loops/_fast_numpy_loops.cpp(395): warning C4244: 'initializing': conversion from 'const int64_t' to 'const npy_intp', possible loss of data
  src/fast_numpy_loops/_fast_numpy_loops.cpp(479): warning C4244: 'argument': conversion from 'int64_t' to 'size_t', possible loss of data
  src/fast_numpy_loops/_fast_numpy_loops.cpp(479): warning C4244: 'argument': conversion from 'int64_t' to 'size_t', possible loss of data
  src/fast_numpy_loops/_fast_numpy_loops.cpp(527): warning C4244: '=': conversion from 'int64_t' to 'npy_intp', possible loss of data
  src/fast_numpy_loops/_fast_numpy_loops.cpp(530): warning C4244: '=': conversion from 'int64_t' to 'npy_intp', possible loss of data
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.27.29110\\bin\\HostX86\\x86\\cl.exe' failed with exit status 2
  ----------------------------------------
  ERROR: Failed building wheel for fast-numpy-loops
ERROR: Failed to build one or more wheels
WARNING: You are using pip version 20.2.3; however, version 20.2.4 is available.
You should consider upgrading via the 'c:\cibw\python\pythonx86.3.6.8\tools\python.exe -m pip install --upgrade pip' command.
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\cibuildwheel\__main__.py", line 319, in <module>
    main()
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\cibuildwheel\__main__.py", line 236, in main
    cibuildwheel.windows.build(build_options)
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\cibuildwheel\windows.py", line 240, in build
    ], env=env)
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\cibuildwheel\windows.py", line 26, in call
    subprocess.check_call([str(a) for a in args], env=env, cwd=cwd, shell=True)
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['pip', 'wheel', 'D:\\a\\numpy-threading-extensions\\numpy-threading-extensions', '-w', 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\cibuildwheeljttxuuoh\\built_wheel', '--no-deps']' returned non-zero exit status 1.
Error: Process completed with exit code 1.
```
@mattip can you see why this is failing?

**Comment by mattip** [comment]
It’s an incompatibility between atop and windows. @tdimitri , thoughts?

**Comment by mattip** [comment]
I added a win32 build and it passes. I think gh-46 fixed it. Closing, please reopen if you see problems.

**Comment by kcpevey** [comment]
Still failing [in CI](https://github.com/Quansight/numpy-threading-extensions/runs/1322680170?check_suite_focus=true). @mattip did you test inside the CI? I wonder what is different?

**Comment by mattip** [comment]
I wonder what is up with tox, it should be using the 32-bit windows python to test with. Maybe we should go back to plain `python setup build --in-place; python -mpytest .` instead of tox?

**Comment by tdimitri** [comment]
might have fixed it.. at least it should compile now

**Comment by mattip** [comment]
xref PR gh-102 to simplify testing. I don't think tox was even running our tests.

**Comment by mattip** [comment]
The test that crashes is [logical_and](https://github.com/Quansight/numpy-threading-extensions/runs/1558727682#step:6:171)

**Comment by mattip** [comment]
@tdimitri did you fix this in gh-106?

**Comment by tdimitri** [comment]
@mattip I dont know.  I saw a minor problem with logical_and (related to the signature) and I cleaned it up.  I don't know why win32 crashed - I would prefer to reproduce the crash and see it in a debugger.  I looked at the logical_and code and... the only thing I could see possible wrong was related to loadu vs loada (unaligned vs aligned).
It might be that the 32bit compiler has different intrinsic optimizations and we have to force loadu.
